src/main/java/store/service/impl/ProductServiceImpl.java

```diff
@@ -43,4 +41,13 @@ public ProductDTO addProduct(ProductDTO productDTO) {
         Product product = mapper.toEntity(productDTO, Product.class);
         return mapper.toDTO(productRepository.save(product), ProductDTO.class);
     }
+
+    @Override
+    public Product checkProductAndUpdateRating(Long idProduct, double rating) {
+        Product product = productRepository.findById(idProduct).orElseThrow(() -> new ProductNotFoundException(idProduct));
+        double oldRating = product.getRating();
+        rating = oldRating == 0 ? rating : ((oldRating + rating) / 2);
```

Comment:
i would recalculate full rating.
load all reviews and calculate median
or you can save rating and reviews amount in product. I prefer last option